ckanext/security/authenticator.py

```
@@ -97,12 +104,19 @@ def authenticate(identity):
     # if the CKAN authenticator has successfully authenticated
     # the request and the user wasn't locked out above,
     # then check the TOTP parameter to see if it is valid
-    if ckan_auth_result is not None:
-        totp_success = authenticate_totp(user_name)
-        # if TOTP was successful -- reset the log in throttle
-        if totp_success:
-            throttle.reset()
-            return ckan_auth_result
+    totp_success = authenticate_totp(user_name)
+    # if TOTP was successful -- reset the log in throttle
+    if totp_success:
+        throttle.reset()
+        return ckan_auth_result
+    else:
+        # This means that the login form has been submitted
+        # with an invalid TOTP code, bypassing the ajax
+        # login() workflow in utils.login.
+        # The username and password were fine, but the 2fa
+        # code was missing or invalid
+        throttle.increment()
+        abort(403, 'Two factor authentication failed, please try again')
```

Just two questions re this error handling:
1. Shouldn't the status be 401 given the authentication flow failed?
2. Is there any risk of letting the user know that the 2FA failed instead of returning generic authentication failed message?

---

1. Yes, you're right
2. Seems unlikely, but I agree that a more generic message would be better

---

I ended up hooking in to the CKAN auth flow at a higher level in order to disguise the login failure modes further so that malicious users will always see the same failure message.